### PR TITLE
allow php8 attributes for Tree extension

### PR DESCRIPTION
**Pull request by tacman**
This PR follows the pattern in Loggable to allow for PHP 8 attributes, #2363

However, it doesn't quite work yet, when trying to use a class that implements this I get

    Tree object class: App\Entity\Location must have tree metadata at this point

I imagine that in src/Tree/Mapping/Driver/Annotation.php, it is not loading the attributes, but I can't figure out where that's happening.  Can someone point me in the right direction?

### CODE REVIEW

**Comment by tacman** [comment]
I added the Attribute class, and am doing some basic testing now.

File: src/Mapping/Annotation/Tree.php [hunk 1]
```diff
@@ -42,4 +45,21 @@ final class Tree extends Annotation
      * @deprecated to be removed in 4.0, unused, configure the property on the TreeRoot annotation instead
      */
     public $identifierMethod;
+
+    /**
+     * @phpstan-param class-string|null $type
+     */
+    public function __construct(array $data = [], ?string $type = null)
```
Comment (franmomu) [inline src/Mapping/Annotation/Tree.php:52]: All the public fields should be in the constructor, see

https://github.com/doctrine-extensions/DoctrineExtensions/blob/81458019f09ae999c4509e670f64594c6d8f7e3c/src/Mapping/Annotation/Timestampable.php#L26-L48

**Review by franmomu — COMMENTED** [review]
thanks for starting this PR!

**Comment by tacman** [comment]
Thanks, I've fixed that and pushed a change.

My tests are failing, but they seem to be related to Mongo.  I don't have
mongo (or the driver) installed locally, but the github actions failure
seem to have nothing to do with my changes.

**Comment by tacman** [comment]
I'm stuck. :-(

It _appears_ that the failures have something to do with mongodb.

File: src/Mapping/Annotation/Slug.php [hunk 2]
```diff
@@ -41,4 +44,38 @@ final class Slug extends Annotation
...
+
+    /**
+     * @phpstan-param class-string|null $type
+     */
+    public function __construct(array $data = [],
+                                array $fields = [],
+                                bool $updatable = true,
+                                string $style = 'default', // or "camel"
+                                bool $unique = true,
+                                string $unique_base = '',
```
Comment (franmomu) [inline src/Mapping/Annotation/Slug.php:56]:
```suggestion
                                ?string $unique_base = null,
```

**Comment by franmomu** [comment]
> It _appears_ that the failures have something to do with mongodb.

Looking at the errors:

https://github.com/doctrine-extensions/DoctrineExtensions/runs/4459312433?check_suite_focus=true#step:8:34

it is not only mongodb, and looking at the line

https://github.com/doctrine-extensions/DoctrineExtensions/blob/81458019f09ae999c4509e670f64594c6d8f7e3c/src/Sluggable/SluggableListener.php#L431

something was not right with the `unique_base` (the phpdoc is not right, it should be `string|null`)

**Review by franmomu — CHANGES_REQUESTED** [review]
thanks @tacman! but there is still some work needed, basically testing.

We need to add support for these attributes in our models like:

https://github.com/doctrine-extensions/DoctrineExtensions/blob/05e0780273389764dc6dcebfcea34bbac84ff6f0/src/Tree/Traits/NestedSetEntity.php#L23-L29

And then, for testing purposes we have to modify the tests in https://github.com/doctrine-extensions/DoctrineExtensions/tree/main/tests/Gedmo/Tree and replace the calls:

`$this->getMockSqliteEntityManager($evm)` -> `$this->getDefaultMockSqliteEntityManager($evm)`
`$this->getMockDocumentManager($evm)` -> `$this->getDefaultDocumentManager($evm)`

This will use AttributeDriver when using PHP >= 8

And for fully test it we should add attributes to the fixtures: https://github.com/doctrine-extensions/DoctrineExtensions/tree/main/tests/Gedmo/Tree/Fixture. This is quite some work that maybe could be done using [rector](https://getrector.org/blog/how-to-upgrade-annotations-to-attributes) (not removing the existing annotation), but I haven't tried it (so far, I've always done it by hand because I was also adding types).

File: .github/workflows/coding-standards.yml [hunk 3]
```diff
@@ -4,6 +4,7 @@ on:
   push:
     branches:
       - main
+      - support-attributes
```
Comment (franmomu) [inline .github/workflows/coding-standards.yml:7]: Please remove this line

**Comment by tacman** [comment]
Sure.  Question -- how do I trigger github actions on my branch?  Or should
I be making changes on my fork's main branch?

**Comment by franmomu** [comment]
> Sure. Question -- how do I trigger github actions on my branch? Or should I be making changes on my fork's main branch?

hmmm I don't know if you can see the actions here: https://github.com/tacman/DoctrineExtensions/actions or you can create also a PR in you fork

**Comment by franmomu** [comment]
> > Sure. Question -- how do I trigger github actions on my branch? Or should I be making changes on my fork's main branch?
> 
> hmmm I don't know if you can see the actions here: https://github.com/tacman/DoctrineExtensions/actions or you can also create a PR in you fork

**Comment by franmomu** [comment]
@tacman if you rebase with the `main` branch, you _only_ have to add the attributes of the Tree extension to tests, the mapping is already done.

**Comment by franmomu** [comment]
@tacman if you give permissions to push in your repository I'll try to work on this

**Comment by franmomu** [comment]
Thanks @tacman for starting this PR! I've used your code and added the missing parts in https://github.com/doctrine-extensions/DoctrineExtensions/pull/2388

**Pull request closed without merging**